File: scripts_matlab/dydt_KDEB.m
function output = dydt_KDEB(t, EVHR, p)
global aux

%% Unpack state vars
E  = EVHR(1); % J, reserve energy
V  = EVHR(2); % cm^3, structural volume
E_H  = EVHR(3); % J , cumulated energy inversted into maturity 
E_R  = EVHR(4); % J, reproduction buffer

vars_pull(p);

%% Metamorphosis check (set acceleration factor)
 if E_H <= E_Hb
    s_M = 1; % acceleration factor before and at birth;
 elseif E_H > E_Hb && E_H < E_Hj
    s_M = V^(1/3)/L_b; % acceleration factor between birth and metamorposis;
 end

%% Temperature correction function (details in Kooijman 2010, p. 21)
%  
 if ismember('tT',fieldnames(aux))
    T = spline1(t, aux.tT);      % use when have temperature vector in time
 else
    T = aux.T + 273.15;
 end

 if ismember('tf',fieldnames(aux))
    f = spline1(t, aux.tf);       % use when have feeding vector in time
 else
    f = aux.f;
 end
 c_T = exp(T_A/ T_ref - T_A/ T);

 p_AmT = c_T * p_Am * s_M;
 v_T = c_T * v * s_M;
 p_MT = c_T * p_M;
 p_TT = c_T * p_T;
 k_JT = c_T * k_J; 
 p_XmT = p_AmT / kap_X;

%% Energy fluxes

% Feeding
 if E_H < E_Hb
    pX = 0; % embryo stage -> f=0
 else
    pX = f * p_XmT * V^(2/3);
 end

% Fluxes
 pA = kap_X * pX;
 pM = p_MT * V;
 pT = p_TT * V^(2/3);
 pS = pM + pT;
 pC = (E/V) * (E_G * v_T * V^(2/3) + pS ) / (kap * E/V + E_G); %eq. 2.12 p.37 Kooijman 2010
 pJ = k_JT * E_H;

% Energy amount check - is there enough in the reseve for maintenance?
 if pA < (kap * pC) && (kap * pC) < pS && E < pS
    output = [-E; 0; 0; 0];
    return
 end
 
%% Differential equations for state variables
 dE = pA - pC;  % dE/dt
 dV = max (0, (kap * pC - pS) / E_G); % dV/dt, max function does not allow shrinking (remove if you want to deplete structure)
 if E_H < E_Hp
    dH = max (0, (1 - kap) * pC - pJ); % dEH/dt
    dR = 0; % dER/dt
 else
    dH = 0;
    dR = max (0, (1 - kap) * pC - pJ); % dR/dt, max function does not allow shrinking of repro buffer (remove if you want to deplete repro buffer)
 end
 
output = [dE; dV; dH; dR];

end

File: deb/dydt_KDEB.py
import numpy as np
# from spline1 import spline1
from scipy.interpolate import interp1d


def dydt_KDEB(t, EVHR, aux, p, EVHR_):
    E = EVHR[0]  # J, reserve energy
    V = EVHR[1]  # cm^3, structural volume
    E_H = EVHR[2]  # J , cumulated energy inversted into maturity
    E_R = EVHR[3]  # J, reproduction buffer

    if E_H <= p["E_Hb"]:
        s_M = 1
    elif E_H > p["E_Hb"] and E_H < p["E_Hj"]:
        s_M = V**(1/3)/p["L_b"]
    else:
        s_M = p["s_M"]

    if "tT" in aux:
        # T = spline1(t, aux["tT"])
        Tf = interp1d(aux["tT"][:, 0], aux["tT"][:,1], kind="slinear")
        T = Tf(t)
    else:
        T = aux["T"] + 273.15
    # print(T, "TTTTTTTTTTTT")

    if "tf" in aux:
        # f = spline1(t, aux["tf"])
        ff = interp1d(aux["tf"][:, 0], aux["tf"][:,1], kind="slinear")
        f = ff(t)
    else:
        f = aux["f"]

    # print(p["T_A"], T)
    c_T = np.exp(p["T_A"]/p["T_ref"] - p["T_A"]/T)

    p_AmT = c_T * p["p_Am"] * s_M
    v_T = c_T * p["v"] * s_M
    p_MT = c_T * p["p_M"]
    p_TT = c_T * p["p_T"]
    k_JT = c_T * p["k_J"]
    p_XmT = p_AmT / p["kap_X"]

    if E_H < p["E_Hb"]:
        pX = 0 # embryo stage -> f=0
    else:
        pX = f * p_XmT * V**(2/3)

    pA = p["kap_X"] * pX
    pM = p_MT * V
    pT = p_TT * V**(2/3)
    pS = pM + pT
    # eq. 2.12 p.37 Kooijman 2010
    pC = (E/V) * (p["E_G"] * v_T * V**(2/3) + pS ) / (p["kap"] * E/V + p["E_G"]) 
    pJ = k_JT * E_H

    if pA < (p["kap"]*pC) and (p["kap"] * pC) <pS and E<pS:
        output = [-E, 0, 0, 0]
        return output
    else:

        dE = pA-pC
        dV = np.max((0, (p["kap"]*pC -pS)/p["E_G"]))
        if E_H < p["E_Hp"]:
            dH = np.max((0, (1-p["kap"])* pC-pJ))
            dR = 0
        else:
            dH = 0
            dR = np.max((0, (1-p["kap"])*pC - pJ))

        output = [dE, dV, dH, dR]

        return output